src/functions.rs

```diff
@@ -1849,13 +1859,37 @@ fn validate_cast(
     {
         return Err(err_ret(start_type.get_range()));
     }
-    match (cast_ti, func_call) {
-        (Some(cast_type_info), Some(func_call_info)) => {
-            // TODO add additional checks.  Also add cases which we will allow casting even if the casting type isn't a parent.
-            validate_inheritance(func_call_info, type_info, &cast_type_info.name, file)
+    match (cast_ti, func_call, func_info) {
+        (Some(cast_ti), Some(func_call), Some(func_info)) => {
+            if !validate_inheritance(func_call, type_info, &cast_ti.name, file)? {
+                if func_info.is_castable {
+                    Ok(())
+                } else {
+                    Err(ErrorItem::make_compile_or_internal_error(
+                        "Not something we can cast to",
+                        file,
+                        func_call.get_name_range(),
+                        "The function is not castable or inherited by caller",
+                    ))
+                }
+            } else {
+                Ok(())
+            }
         }
-        (None, _) => Err(err_ret(start_type.get_range())),
-        (_, _) => Ok(()),
+        (Some(_), Some(_), None) => Err(ErrorItem::make_compile_or_internal_error(
```

Comment:
We don't want to expose the term "Function Info" to users.  That's an internal code term that policy developers won't understand.

When can this happen?  It looks to me like the current callers either pass Some(_), Some(_) or None, None for the call and the info.  I think it would just be a compiler bug for them to pass us a call, but no info.  We should probably just document that it's a requirement to be consistent there in the comments for this function and return an InternalError in this case.